### PR TITLE
fix(autoware_ground_segmentation): fix scan ground filter logic 

### PR DESCRIPTION
**Pull request by technolojin**
## Description
Fixing the scan ground filter logic implementation to match original algorithm.

1. `classifyPointCloudGridScan`
    1. Initialization process is refactored
    2. Revise condition to finalize the first grid filling
2. `initializeFirstGndGrids`
    1. Refactored for readability
3. `checkContinuousGndGrid`
    1. A bug that calculate the mean position of the grids is fixed
    2. Determines the local slope condition earlier for efficiency
4. `checkDiscontinuousGndGrid`, `checkBreakGndGrid`
    1.  Determines each condition and return for efficiency

## Related links

**Parent Issue:**

- Link

## How was this PR tested?

[TIER IV Cloud](https://evaluation.tier4.jp/evaluation/reports/3217febb-a242-5fb8-b050-f23709638dc6?project_id=prd_jt)

## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

File: perception/autoware_ground_segmentation/src/scan_ground_filter/node.hpp [hunk 1]
```diff
@@ -175,21 +175,21 @@ class ScanGroundFilterComponent : public autoware::pointcloud_preprocessor::Filt
...
-  double global_slope_max_ratio_;
-  double local_slope_max_ratio_;
-  double split_points_distance_tolerance_;  // distance in meters between concentric divisions
-  double split_points_distance_tolerance_square_;
+  float global_slope_max_angle_rad_;  // radians
+  float local_slope_max_angle_rad_;   // radians
+  float global_slope_max_ratio_;
+  float local_slope_max_ratio_;
+  float split_points_distance_tolerance_;  // distance in meters between concentric divisions
+  float split_points_distance_tolerance_square_;
```
Comment (badai-nguyen) [inline perception/autoware_ground_segmentation/src/scan_ground_filter/node.hpp:188]: I think the primary types of paramters getting through declare_parameter should be double as [here](https://docs.ros.org/en/foxy/Concepts/About-ROS-2-Parameters.html#overview)

File: perception/autoware_ground_segmentation/src/scan_ground_filter/node.cpp [hunk 2]
```diff
@@ -270,83 +282,107 @@ void ScanGroundFilterComponent::checkContinuousGndGrid(
...
+  const float gnd_z_local_thresh =
+    std::tan(DEG2RAD(5.0)) * (pd.radius - gnd_grids_list.back().radius);
 
-  tmp_delta_mean_z = point_curr.z - (gnd_grids_list.end() - 2)->avg_height;
-  tmp_delta_radius = pd.radius - (gnd_grids_list.end() - 2)->radius;
-  float local_slope_ratio = tmp_delta_mean_z / tmp_delta_radius;
-  if (
-    abs(point_curr.z - next_gnd_z) <= non_ground_height_threshold_ + gnd_z_local_thresh ||
-    abs(local_slope_ratio) <= local_slope_max_ratio_) {
+  if (abs(point_curr.z - next_gnd_z) < non_ground_height_threshold_ + gnd_z_local_thresh) {
```
[the remark this answers is not in the document]
Comment (technolojin) [reply perception/autoware_ground_segmentation/src/scan_ground_filter/node.cpp]: I think it was a mistake.
Fixed https://github.com/autowarefoundation/autoware.universe/pull/9084/commits/09d8a5a438c8a983acd6f104b8a0da6e04148a90

File: perception/autoware_ground_segmentation/src/scan_ground_filter/node.hpp (hunk 1 above)
Comment (technolojin) [reply perception/autoware_ground_segmentation/src/scan_ground_filter/node.hpp:188]: You also set other parameters in float. Shall I align all the parameters in float?

**Review by badai-nguyen — APPROVED** [review]
LGTM

**Pull request merged**